tikv/raftstore/worker.go

```diff
@@ -290,7 +291,75 @@ type pendingDeleteRanges struct {
 }
 
 func (pendDelRanges *pendingDeleteRanges) insert(regionId uint64, startKey, endKey []byte, timeout time.Time) {
-	// todo, currently it is a place holder.
+	if len(pendDelRanges.findOverlapRanges(startKey, endKey)) != 0 {
+		panic(fmt.Sprintf("[region %d] register deleting data in [%v, %v) failed due to overlap", regionId, startKey, endKey))
+	}
+	peerInfo := &stalePeerInfo{
+		endKey:   endKey,
+		regionId: regionId,
+		timeout:  timeout,
+	}
+	pendDelRanges.ranges.Insert(startKey, peerInfo.serializeStalePeerInfo())
+}
+
+// remove removes and returns the peer info with the `start_key`.
+func (pendDelRanges *pendingDeleteRanges) remove(startKey []byte) *stalePeerInfo {
+	value := pendDelRanges.ranges.Get(startKey, nil)
+	if value != nil {
+		pendDelRanges.ranges.Delete(startKey)
+		return deserializeStalePeerInfo(safeCopy(value))
+	}
+	return nil
+}
+
+// timeoutRanges returns all timeout ranges info.
+func (pendDelRanges *pendingDeleteRanges) timeoutRanges(now time.Time) (ranges []delRangeHolder) {
+	ite := pendDelRanges.ranges.NewIterator()
+	for ite.Next(); ite.Valid(); ite.Next() {
+		startKey := safeCopy(ite.Key())
+		peerInfo := deserializeStalePeerInfo(safeCopy(ite.Value()))
+		if peerInfo.timeout.Before(now) {
+			ranges = append(ranges, delRangeHolder{
+				startKey: startKey,
+				endKey:   peerInfo.endKey,
+				regionId: peerInfo.regionId,
+			})
+		}
+	}
+	return
+}
+
+type stalePeerInfo struct {
+	regionId uint64
+	endKey   []byte
+	timeout  time.Time
+}
+
+func (peerInfo *stalePeerInfo) serializeStalePeerInfo() []byte {
+	if timeoutData, err := peerInfo.timeout.MarshalBinary(); err != nil {
+		log.Errorf("serialize stale Peer info failed, [regionId: %d, endKey: %v]", peerInfo.regionId, peerInfo.endKey)
+		return nil
+	} else {
+		buf := make([]byte, 8+4+4)
+		binary.LittleEndian.PutUint64(buf, peerInfo.regionId)
+		binary.LittleEndian.PutUint32(buf[8:], uint32(len(peerInfo.endKey)))
+		binary.LittleEndian.PutUint32(buf[12:], uint32(len(timeoutData)))
```

Comment:
We can just store a single int64 for time.
```
        t := time.Now()
	n := t.UnixNano()
	t = time.Unix(0, n)
```
And we also don't need to store the length of endkey, if the total length of value is known, and other two fields takes 16 bytes.